### PR TITLE
Merge GitLab merge requests when shipping

### LINKED ISSUE #1092: Shipping a branch does not close GitLab MR
When using `git town ship` with a GitLab repo, the MR is not merged and the branch is deleted.
The MR stays opened because it can't be merged (the code has already been merged but the MR doesn't know it) and it can be closed but that's not what really happened.

Squash MR is not available for GitLab. They provide a [Squash and Merge](https://gitlab.com/help/user/project/merge_requests/squash_and_merge) feature on the Web UI and it is only for GitLab EES.  All the commits are squashed in one and then a non fast forward merge is applied on the remaining commit.

I've also looked up inside [GitLab API](https://docs.gitlab.com/ce/api/merge_requests.html#accept-mr), it is actually not possible to squash MR throught API.

Do you have any idea to update the MR status to merged with git town ship?

Thanks

### PR DESCRIPTION
**Pull request by manuelprinz**
Hi, this is my attempt to fix #1092. It is my first real contact with Go (except from some toying around), so some of it might look weird or not be best practice. I appreciate any feedback!

I had some issues with getting the test to run. There seemed to be an issue with some commit messages. I included a "fix" in this branch. If it is just on my box, I'm happy to revert it.

I used the GitHub driver as an example. The inner working is pretty much the same, so there is quite a bit of duplication, also in the tests. I had to shut up the linter, and am fully aware that this is not the right solution, but I did not know any better way for the time being. (As I said, Go beginner.) I guess that having something like a template method might be a solution, but some guidance would be appreciated.

It also worked when testing manually. I have not tested it with [all the merge strategies in GitLab](https://docs.gitlab.com/ee/user/project/merge_requests/methods/index.html), however. It seems to work fine with "Merge commit" and "Merge commit with semi-linear history". I would expect some issues with "Fast forward", but I have not tried.

Anyway, thanks for writing Git Town! I really love it!

Best, Manuel

### CODE REVIEW

**Review by kevgo — COMMENTED** [review]
This PR is very well done. I really only have a few nitpicks. Sorry for the trouble with local testing and thanks for adding this important feature!

**Comment by kevgo** [comment]
Please sync this branch to fix the CI build.

File: features/sync/current_branch/feature_branch/features/pull_branch_strategy.feature [hunk 1]
```diff
@@ -28,11 +28,11 @@ Feature: with pull-branch-strategy set to "merge"
       | BRANCH  | LOCATION      | MESSAGE                                                    |
       | main    | local, origin | local main commit                                          |
       |         |               | origin main commit                                         |
-      |         |               | Merge remote-tracking branch 'origin/main'                 |
+      |         |               | Merge remote-tracking branch 'origin/main' into main       |
```
[the remark this answers is not in the document]
Comment (manuelprinz) [reply features/sync/current_branch/feature_branch/features/pull_branch_strategy.feature:31]: You are right, that box has 2.20.1. I'll revert the commit.

File: src/hosting/gitlab.go [hunk 2]
```diff
@@ -46,14 +76,101 @@ func (d *GitlabDriver) NewPullRequestURL(branch, parentBranch string) (string, e
 	return fmt.Sprintf("%s/merge_requests/new?%s", d.RepositoryURL(), query.Encode()), nil
 }
 
+func (d *GitlabDriver) BaseURL() string {
+	return fmt.Sprintf("https://%s", d.hostname)
+}
```
[the remark this answers is not in the document]
Comment (manuelprinz) [reply src/hosting/gitlab.go:81]: Thanks!

File: src/hosting/gitlab.go [hunk 3]
```diff
@@ -28,14 +34,38 @@ func NewGitlabDriver(config config) *GitlabDriver {
...
+	if d.apiToken == "" {
+		return result, nil
+	}
+	d.connect()
+	mergeRequests, err := d.loadMergeRequests(branch, parentBranch)
+	if err != nil {
+		return result, err
+	}
+	if len(mergeRequests) != 1 {
+		return result, err
```
[the remark this answers is not in the document]
Comment (manuelprinz) [reply src/hosting/gitlab.go:64]: Added.

**Comment by manuelprinz** [comment]
> This PR is very well done. I really only have a few nitpicks. Sorry for the trouble with local testing and thanks for adding this important feature!

Thank you for your kind words and quick reply!

**Review by kevgo — APPROVED** [review]

**Comment by kevgo** [comment]
Thanks, this is ready to go. We typically update code and documentation together (thanks to this being a mono-repo). Can you please mention this new GitLab support on https://github.com/git-town/git-town/blob/main/website/src/commands/ship.md and https://github.com/git-town/git-town/blob/main/website/src/quick-configuration.md#api-access-to-your-hosting-provider. Sorry that I didn't mention this earlier.

**Comment by manuelprinz** [comment]
Sure, no problem! I was away for a couple of days and will update the MR during the next days.

**Comment by manuelprinz** [comment]
I updated the documentation. Hope this is what you had in mind! I also put in some GitLab-specific doc because it might confuse GitLab users. I added it to the configuration section, as I did not find a better spot. Please let me know if I should change that!

**Comment by kevgo** [comment]
Thanks for contributing this awesome feature! 🚀

**Pull request merged**